File: convert_to_txt.py
#!/usr/bin/env python3
"""
Generate .txt files from markdown (default flow) or, optionally, from .docx by
first converting them to markdown in a temporary folder.
"""

import argparse
import sys
import tempfile
from pathlib import Path

BASE_DIR = Path(__file__).parent
MD_DIR = BASE_DIR / "Templates_markdown"
DOCX_DIR = BASE_DIR / "Templates_docx"
TXT_DIR = BASE_DIR / "Templates_txt"


def clean_markdown_text(text: str) -> str:
    """Remove basic markdown markers (* and #) while keeping line breaks."""
    return text.replace("*", "").replace("#", "")


def convert_md_file(md_path: Path, output_dir: Path) -> None:
    """Convert a markdown file to txt, stripping markers."""
    output_dir.mkdir(parents=True, exist_ok=True)
    txt_path = output_dir / f"{md_path.stem}.txt"
    content = md_path.read_text(encoding="utf-8")
    cleaned = clean_markdown_text(content)
    txt_path.write_text(cleaned, encoding="utf-8")


def convert_markdown_folder(md_dir: Path, output_dir: Path) -> None:
    """Convert all .md files in a folder."""
    md_files = sorted(md_dir.glob("*.md"))
    if not md_files:
        print(f"No .md files found in {md_dir}")
        return

    for md_file in md_files:
        convert_md_file(md_file, output_dir)
        print(f"✓ {md_file.name} -> {md_file.stem}.txt")


def convert_from_docx(output_dir: Path) -> None:
    """Alternate flow: convert .docx to temporary markdown, then to txt."""
    try:
        from convert_to_markdown import convert_docx_to_markdown
    except Exception as exc:  # pragma: no cover - defensive import
        print("Erro ao importar convert_docx_to_markdown de convert_to_markdown.py")
        raise SystemExit(exc)

    docx_files = sorted(DOCX_DIR.glob("*.docx"))
    if not docx_files:
        print(f"No .docx files found in {DOCX_DIR}")
        return

    with tempfile.TemporaryDirectory(prefix="md_tmp_") as tmp_dir:
        tmp_md_dir = Path(tmp_dir)
        for docx_file in docx_files:
            markdown_content = convert_docx_to_markdown(docx_file)
            md_output = tmp_md_dir / f"{docx_file.stem}.md"
            md_output.write_text(markdown_content, encoding="utf-8")
            print(f"Generated temporary: {md_output.name}")

        convert_markdown_folder(tmp_md_dir, output_dir)
        # TemporaryDirectory se encarrega de apagar tmp_md_dir


def parse_args(argv=None):
    parser = argparse.ArgumentParser(
        description=(
            "Converts markdown -> txt (default). "
            "Use --from-docx to convert docx -> markdown -> txt."
        )
    )
    parser.add_argument(
        "--from-docx",
        action="store_true",
        help="Generate txt from .docx (with an intermediate markdown conversion).",
    )
    return parser.parse_args(argv)


def main(argv=None) -> None:
    args = parse_args(argv)

    if args.from_docx:
        convert_from_docx(TXT_DIR)
    else:
        if not MD_DIR.exists():
            raise SystemExit(f"Source folder not found: {MD_DIR}")
        convert_markdown_folder(MD_DIR, TXT_DIR)

    print(f"\n✓ Arquivos gerados em {TXT_DIR}")


if __name__ == "__main__":
    main(sys.argv[1:])

File: rust_converters/src/bin/convert_to_txt.rs
use anyhow::Result;
use docx_rust::document::{BodyContent, Paragraph, ParagraphContent};
use docx_rust::DocxFile;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use tempfile::TempDir;

fn clean_markdown_text(text: &str) -> String {
    text.replace('*', "").replace('#', "")
}

fn convert_md_file(md_path: &Path, output_dir: &Path) -> Result<()> {
    fs::create_dir_all(output_dir)?;
    let txt_path = output_dir.join(
        md_path
            .file_stem()
            .unwrap()
            .to_string_lossy()
            .to_string()
            + ".txt",
    );
    let content = fs::read_to_string(md_path)?;
    let cleaned = clean_markdown_text(&content);
    fs::write(txt_path, cleaned)?;
    Ok(())
}

fn convert_markdown_folder(md_dir: &Path, output_dir: &Path) -> Result<()> {
    let mut md_files: Vec<PathBuf> = fs::read_dir(md_dir)?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().and_then(|s| s.to_str()) == Some("md"))
        .collect();

    if md_files.is_empty() {
        println!("No .md files found in {}", md_dir.display());
        return Ok(());
    }

    md_files.sort();
    for md_file in md_files {
        convert_md_file(&md_file, output_dir)?;
        println!(
            "✓ {} -> {}.txt",
            md_file.file_name().unwrap().to_string_lossy(),
            md_file
                .file_stem()
                .unwrap()
                .to_string_lossy()
        );
    }

    Ok(())
}

fn paragraph_to_markdown(p: &Paragraph) -> String {
    let plain = p.text();
    if plain.trim().is_empty() {
        return String::new();
    }

    let mut text_parts: Vec<String> = Vec::new();
    for pc in &p.content {
        if let ParagraphContent::Run(run) = pc {
            let mut text = run.text();
            if text.is_empty() {
                continue;
            }
            if let Some(prop) = &run.property {
                if prop.bold.is_some() {
                    text = format!("**{}**", text);
                }
                if prop.italics.is_some() {
                    text = format!("*{}*", text);
                }
                if prop.underline.is_some() {
                    text = format!("__{}__", text);
                }
            }
            text_parts.push(text);
        }
    }

    if text_parts.is_empty() {
        plain
    } else {
        text_parts.join("")
    }
}

fn convert_docx_to_markdown(docx_path: &Path) -> Result<String> {
    let file = DocxFile::from_file(docx_path)?;
    let docx = file.parse()?;

    let mut markdown_lines: Vec<String> = Vec::new();
    let body = &docx.document.body;

    for item in &body.content {
        if let BodyContent::Paragraph(p) = item {
            markdown_lines.push(paragraph_to_markdown(p));
        }
    }

    Ok(markdown_lines.join("\n"))
}

fn convert_from_docx(output_dir: &Path) -> Result<()> {
    let docx_dir = PathBuf::from("Templates_docx");
    let mut docx_files: Vec<PathBuf> = fs::read_dir(&docx_dir)?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().and_then(|s| s.to_str()) == Some("docx"))
        .collect();

    if docx_files.is_empty() {
        println!("No .docx files found in {}", docx_dir.display());
        return Ok(());
    }

    docx_files.sort();

    let tmp_dir: TempDir = TempDir::new()?;
    let tmp_md_dir = tmp_dir.path();

    for docx_file in &docx_files {
        let markdown_content = convert_docx_to_markdown(docx_file)?;
        let md_output = tmp_md_dir.join(
            docx_file
                .file_stem()
                .unwrap()
                .to_string_lossy()
                .to_string()
                + ".md",
        );
        fs::write(&md_output, markdown_content)?;
        println!(
            "Generated temporary: {}",
            md_output.file_name().unwrap().to_string_lossy()
        );
    }

    convert_markdown_folder(tmp_md_dir, output_dir)?;
    // TempDir cleans up automatically when it goes out of scope
    Ok(())
}

fn main() -> Result<()> {
    let args: Vec<String> = env::args().skip(1).collect();
    let from_docx = args.iter().any(|a| a == "--from-docx");

    let md_dir = PathBuf::from("Templates_markdown");
    let txt_dir = PathBuf::from("Templates_txt");

    if from_docx {
        convert_from_docx(&txt_dir)?;
    } else {
        if !md_dir.exists() {
            anyhow::bail!("Source folder not found: {}", md_dir.display());
        }
        convert_markdown_folder(&md_dir, &txt_dir)?;
    }

    println!("\n✓ Files generated in {}", txt_dir.display());
    Ok(())
}
